src/node/vscode.ts

```diff
@@ -0,0 +1,150 @@
+      vscode.removeAllListeners()
+      this._vscode = undefined
+      vscode.kill()
+    }
+  }
+
+  public async initialize(
+    options: Omit<ipc.VscodeOptions, "startPath">,
+    query: ipc.Query,
+  ): Promise<ipc.WorkbenchOptions> {
+    const { lastVisited } = await settings.read()
+    const startPath = await this.getFirstPath([
+      { url: query.workspace, workspace: true },
+      { url: query.folder, workspace: false },
+      options.args._ && options.args._.length > 0
+        ? { url: path.resolve(options.args._[options.args._.length - 1]) }
+        : undefined,
+      lastVisited,
+    ])
+
+    settings.write({
+      lastVisited: startPath,
+      query,
+    })
+
+    const id = generateUuid()
+    const vscode = await this.fork()
+
+    logger.debug("setting up vs code...")
+    return new Promise<ipc.WorkbenchOptions>((resolve, reject) => {
+      vscode.once("message", (message: ipc.VscodeMessage) => {
+        logger.debug("got message from vs code", field("message", message))
+        return message.type === "options" && message.id === id
+          ? resolve(message.options)
+          : reject(new Error("Unexpected response during initialization"))
+      })
+      vscode.once("error", reject)
+      vscode.once("exit", (code) => reject(new Error(`VS Code exited unexpectedly with code ${code}`)))
+      this.send(
+        {
+          type: "init",
+          id,
+          options: {
+            ...options,
+            startPath,
+          },
+        },
+        vscode,
+      )
+    })
+  }
+
+  private fork(): Promise<cp.ChildProcess> {
+    if (!this._vscode) {
+      logger.debug("forking vs code...")
+      const vscode = cp.fork(path.join(this.serverRootPath, "fork"))
+      vscode.on("error", (error) => {
+        logger.error(error.message)
+        this._vscode = undefined
+      })
+      vscode.on("exit", (code) => {
+        logger.error(`VS Code exited unexpectedly with code ${code}`)
+        this._vscode = undefined
+      })
+
+      this._vscode = new Promise((resolve, reject) => {
+        vscode.once("message", (message: ipc.VscodeMessage) => {
+          logger.debug("got message from vs code", field("message", message))
+          return message.type === "ready"
+            ? resolve(vscode)
+            : reject(new Error("Unexpected response waiting for ready response"))
+        })
+        vscode.once("error", reject)
+        vscode.once("exit", (code) => reject(new Error(`VS Code exited unexpectedly with code ${code}`)))
+      })
+    }
+
+    return this._vscode
+  }
+
+  /**
+   * VS Code expects a raw socket. It will handle all the web socket frames.
+   */
+  public async sendWebsocket(socket: net.Socket, query: ipc.Query): Promise<void> {
+    // TODO: TLS socket proxy.
+    const vscode = await this._vscode
+    this.send({ type: "socket", query }, vscode, socket)
+  }
+
+  private send(message: ipc.CodeServerMessage, vscode?: cp.ChildProcess, socket?: net.Socket): void {
```

Comment:
`vscode` is always present here.

Reply:
It might not be, `sendWebsocket` for example can call `this.send` with `undefined` (if `this._vscode` is undefined). I think I'll move that check to `sendWebsocket` though because it's only used in one other place and it definitely does exist there.

Reply:
Ah well maybe not since then I have two places where it throws `vscode is not running` which feels a little unfortunate.